src/hotspot/share/opto/rangeinference.cpp

```diff
@@ -0,0 +1,773 @@
+#include "utilities/tuple.hpp"
+
+constexpr juint SMALLINT = 3; // a value too insignificant to consider widening
+
+// This represents the result of an iterative calculation
+template <class T>
+class AdjustResult {
+public:
+  bool _progress; // whether there is progress compared to the last iteration
+  bool _present;  // whether the calculation arrives at contradiction
+  T _data;
+
+  static AdjustResult<T> make_empty() {
+    return {true, false, {}};
+  }
+};
+
+// In the canonical form, [lo, hi] intersects with [ulo, uhi] can result in 2
+// cases:
+// - [lo, hi] is the same as [ulo, uhi], lo and hi are both >= 0 or both < 0.
+// - [lo, hi] is not the same as [ulo, uhi], which results in the intersections
+// being [lo, uhi] and [ulo, hi], lo and uhi are < 0 while ulo and hi are >= 0.
+// This class deals with each interval with both bounds being >= 0 or < 0 in
+// the signed domain.
+template <class U>
+class SimpleCanonicalResult {
+  static_assert(std::is_unsigned<U>::value, "bit info should be unsigned");
+public:
+  bool _present; // whether this is an empty set
+  RangeInt<U> _bounds;
+  KnownBits<U> _bits;
+
+  static SimpleCanonicalResult<U> make_empty() {
+    return {false, {}, {}};
+  }
+};
+
+// Find the minimum value that is not less than lo and satisfies bits
+template <class U>
+static U adjust_lo(U lo, const KnownBits<U>& bits) {
+  constexpr size_t W = sizeof(U) * 8;
+  // Violation of lo with respects to bits
+  // E.g: lo    = 1100
+  //      zeros = 0100
+  //      ones  = 1001
+  // zero_violation = 0100, i.e the second bit should be zero, but it is 1 in
+  // lo. Similarly, one_violation = 0001, i.e the LSB should be one, but it is
+  // 0 in lo. These make lo not satisfy the bit constraints, which results in
+  // us having to find the smallest value that satisfies bits
+  U zero_violation = lo & bits._zeros;
+  U one_violation = ~lo & bits._ones;
+  if (zero_violation == one_violation) {
+    // This means lo does not violate bits, it is the result
+    assert(zero_violation == 0, "");
+    return lo;
+  }
+
+  // The principle here is that, consider the first bit in result that is
+  // different from the corresponding bit in lo, since result is larger than lo
+  // the bit must be 0 in lo and 1 in result. As result should be the smallest
+  // value, this bit should be the rightmost one possible.
+  // E.g:      1 2 3 4 5 6
+  //      lo = 1 0 0 1 1 0
+  //       x = 1 0 1 0 1 0
+  //       y = 0 1 1 1 1 1
+  // x would be larger than lo since the first different bit is the 3rd one,
+  // while y is smaller than lo because the first different bit is the 1st bit.
+  // Next consider:
+  //      x1 = 1 0 1 0 1 0
+  //      x2 = 1 0 0 1 1 1
+  // Both x1 and x2 are larger than lo, but x1 > x2 since its first different
+  // bit from lo is the 3rd one, while with x2 it is the 7th one. As a result,
+  // if both x1 and x2 satisfy bits, x2 would be closer to our true result.
+  if (zero_violation < one_violation) {
```

Comment:
Hmm I'm struggling to understand this.
Maybe you first should say that this is a case-distinction over the most-significant bit that has a violation.
Then explain/prove why the check `zero_violation < one_violation` does that trick.